tests/unit/igraph_spanner.c

```diff
@@ -0,0 +1,147 @@
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#include <igraph.h>
+#include "test_utilities.inc"
+#include <stdlib.h>
+
+igraph_bool_t _test_spanner (igraph_t *graph, igraph_t *spanner, double stretch, igraph_vector_t *weights, igraph_vector_t *spanner_weights) {
+    /* check the spanner and compare it to the original graph in several topics:
+        1) Compare the number of nodes.
+        2) If the weights of the edges are equal between the graphs (doesn't work on multigraph).
+        3) the stretch factor of the spanner.
+        returns True or false
+    */
+   long int spanner_no_of_nodes = igraph_vcount(spanner);
+   long int no_of_nodes = igraph_vcount(graph);
+   long int no_of_edges = igraph_ecount(graph);
+   long int no_of_edges_spanner = igraph_ecount(spanner);
+   igraph_integer_t from, to, edge_spanner, edge;
+   igraph_eit_t graph_edges;
+   igraph_matrix_t res_spanner, res_graph;
+
+    // compare number of nodes
+    IGRAPH_ASSERT(spanner_no_of_nodes == no_of_nodes);
+    // compare number of edges
+    IGRAPH_ASSERT(no_of_edges > no_of_edges_spanner);
+
+    igraph_matrix_init(&res_spanner, 0, 0);
+    igraph_matrix_init(&res_graph, 0, 0);
+
+    igraph_eit_create(graph, igraph_ess_all(IGRAPH_EDGEORDER_ID), &graph_edges);
+    // compare the edges and their weight
+    for (IGRAPH_EIT_RESET(graph_edges); !IGRAPH_EIT_END(graph_edges); IGRAPH_EIT_NEXT(graph_edges)) {
+        edge = IGRAPH_EIT_GET(graph_edges);
+        igraph_edge(graph, edge, &from, &to);
+        igraph_get_eid(spanner, &edge_spanner, from, to, IGRAPH_UNDIRECTED, 0);
+        if (edge != -1) {
+            double weight_graph = VECTOR(*weights)[edge];
+            double weight_spanner = VECTOR(*spanner_weights)[edge_spanner];
+            IGRAPH_ASSERT(weight_graph == weight_graph);
+        }
+    }
+
+    // Validate the stretch factor
+    igraph_shortest_paths_dijkstra(graph, &res_graph, igraph_vss_all(), igraph_vss_all(), weights, IGRAPH_ALL);
+    igraph_shortest_paths_dijkstra(spanner, &res_spanner, igraph_vss_all(), igraph_vss_all(), spanner_weights, IGRAPH_ALL);
+    for (int x = 0; x < no_of_nodes; x++) {
+        for (int y = 0; y < no_of_nodes; y++) {
+            if (x == y) {
+                continue;
+            }
+            IGRAPH_ASSERT(MATRIX(res_spanner, x, y) < MATRIX(res_graph, x, y) * stretch);
+        }
+    }
+    igraph_matrix_destroy(&res_graph);
+    igraph_matrix_destroy(&res_spanner);
+
+    return IGRAPH_SUCCESS;
+}
+
+int main () {
+    igraph_t graph, spanner;
+    igraph_vector_t weights, spanner_weight;
+    igraph_bool_t check_spanner;
+    igraph_rng_t rng;
+    long int no_of_edges, no_of_edges_spanner;
+    IGRAPH_CHECK(igraph_rng_init(&rng, &igraph_rngtype_mt19937));
```

Comment:
Please do not use `IGRAPH_CHECK` here. It should only be used in functions that return an igraph error code. In the future the type of igraph error codes is likely to change, and may not be compatible with a program exit code.

(All that said, use the default RNG unless there is a very good reason to create your own RNG.)

Reply:
I've sorted this out.